src/Mvc/src/Microsoft.AspNetCore.Mvc.ViewFeatures/HtmlHelperOfT.cs

```diff
@@ -361,41 +349,72 @@ public string NameFor<TResult>(Expression<Func<TModel, TResult>> expression)
                 throw new ArgumentNullException(nameof(expression));
             }
 
-            var modelExplorer = GetModelExplorer(expression);
+            var modelExpression = GetModelExpression(expression);
             return GenerateTextBox(
-                modelExplorer,
-                GetExpressionName(expression),
-                modelExplorer.Model,
+                modelExpression.ModelExplorer,
+                modelExpression.Name,
+                modelExpression.Model,
                 format,
                 htmlAttributes);
         }
 
-        protected string GetExpressionName<TResult>(Expression<Func<TModel, TResult>> expression)
+        /// <summary>
+        /// Gets the <see cref="ModelExpression"/> for <paramref name="expression"/>.
+        /// </summary>
+        /// <typeparam name="TResult">The type of the result.</typeparam>
+        /// <param name="expression">The expression.</param>
+        /// <returns>The <see cref="ModelExpression"/>.</returns>
+        public ModelExpression GetModelExpression<TResult>(Expression<Func<TModel, TResult>> expression)
```

Comment:
The intent is to allow users to downcast the `IHtmlHelper<T>` to `HtmlHelper<T>` instances in their custom extension methods and use it.

Reply:
Ick. That's not discoverable at all. `HtmlHelper<T>` was `public` only for subclassing -- a bad idea in retrospect. And, encouraging people to use it in this way doesn't make sense to me.

Why would we ever recommend downcasting and calling this method? That is, getting the `ModelExpressionProvider` (which is `public`) from DI seems like the go-to approach here.